Create topic operation

evtimo opened the pull request.

evtimo committed:
admin client cache

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <timotshuk.ev@edu.spbstu.ru>

evtimo committed:
topic create operation

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <timotshuk.ev@edu.spbstu.ru>

evtimo committed:
complete operation

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <timotshuk.ev@edu.spbstu.ru>

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -88,7 +113,38 @@ private void submitTopicOperation(PathOperation op, OpType opType) {
-  private void submitTopicCreateOperation() {}
+  private void submitTopicCreateOperation(final PathOperation op) {
+
+    final String topicName = op.dstPath();
```
why not `op.item().name()`?
May topic name contain some path?

evtimo replied:
fixed

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -88,7 +113,38 @@ private void submitTopicOperation(PathOperation op, OpType opType) {
+  private void submitTopicCreateOperation(final PathOperation op) {
+
+    final String topicName = op.dstPath();
+    try {
```
let's use single properties instance (field) and single adminClient instance too

evtimo replied:
will be added in pr #33

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -88,7 +113,38 @@ private void submitTopicOperation(PathOperation op, OpType opType) {
+      val adminClient = adminClientCache.computeIfAbsent("kafka", adminClientCreateFunc);
+      val newTopic = new NewTopic(topicName, 1, (short) 1);
+      final CreateTopicsResult createTopicsResult = adminClient.createTopics(Collections.singleton(newTopic));
+      createTopicsResult.values().get(topicName).get();
```
use a handler for the non-blocking execution

evtimo replied:
fixed. used whenComplete()

akurilov commented on src/main/resources/kafka.properties:
```diff
@@ -0,0 +1,6 @@
+bootstrap.servers=localhost:9092
```
unnecessary file
all this options should be set via the configuration

evtimo replied:
properties file deleted

AnastasiaShemyakinskaya commented on src/main/java/com/emc/mongoose/storage/driver/kafka/cache/AdminClientCreateFunctionImpl.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    public final AdminClient apply(final String name) {
+        properties.setProperty("group.id", name);
```
According to documentation AdminClient doesn't have field "group.id". It has "client.id".

evtimo committed:
add handler & fix

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <timotshuk.ev@edu.spbstu.ru>

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -88,7 +112,41 @@ private void submitTopicOperation(PathOperation op, OpType opType) {
+      val newTopic = new NewTopic(topicName, 1, (short) 1);
+      val createTopicsResult = adminClient.createTopics(Collections.singleton(newTopic));
+      val createTopicFuture = createTopicsResult.values().get(topicName).
+              whenComplete((result, exception) -> {
```
Let's use some driver method reference instead of creating the handler instance every time

evtimo committed:
driver method reference

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <timotshuk.ev@edu.spbstu.ru>

evtimo committed:
Merge branch 'master' of https://github.com/emc-mongoose/mongoose-storage-driver-kafka into KAFKA-34-create-topic-operation

# Conflicts:
#	src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -97,7 +134,29 @@ private void submitTopicOperation(PathOperation op, OpType opType) {
+      val newTopic = new NewTopic(topicName, 1, (short) 1);
+      val createTopicsResult = adminClient.createTopics(Collections.singleton(newTopic));
+      val createTopicFuture = createTopicsResult.values().get(topicName).
+              whenComplete(kafkaTopicFuture(topicName, op));
```
please try using something like this:
`whenComplete((result, thrown) -> handleTopicCreate(op, topicName, thrown));`

evtimo replied:
changed

evtimo committed:
whenComplete

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <timotshuk.ev@edu.spbstu.ru>

akurilov commented:
obsolete
